code/controllers/subsystem/fluids.dm

```diff
@@ -0,0 +1,266 @@
+		currentrun.len--
+
+		if (QDELETED(to_process) || to_process.process(delta_time) == PROCESS_KILL)
+			effect_carousel[cached_bucket_index] -= to_process
+			to_process.effect_bucket = null
+			to_process.datum_flags &= ~DF_ISPROCESSING
+
+		if (MC_TICK_CHECK)
+			break
+
+	if(!currentrun.len)
+		resumed_effect_processing = FALSE
+
+
+/**
+ * Queues a fluid node to spread later after one full carousel rotation.
+ *
+ * Arguments:
+ * - [node][/obj/effect/particle_effect/fluid]: The node to queue to spread.
+ */
+/datum/controller/subsystem/fluids/proc/queue_spread(obj/effect/particle_effect/fluid/node)
+	if (node.spread_bucket)
+		return
+
+	spread_carousel[spread_bucket_index] += node
+	node.spread_bucket = spread_bucket_index
+
+/**
+ * Cancels a queued spread of a fluid node.
+ *
+ * Arguments:
+ * - [node][/obj/effect/particle_effect/fluid]: The node to cancel the spread of.
+ */
+/datum/controller/subsystem/fluids/proc/cancel_spread(obj/effect/particle_effect/fluid/node)
+	if(!node.spread_bucket)
+		return
+
+	var/bucket_index = node.spread_bucket
+	spread_carousel[bucket_index] -= node
+	if (bucket_index == spread_bucket_index)
+		currently_spreading -= node
+
+	node.spread_bucket = null
+
+/**
+ * Starts processing the effects of a fluid node.
+ *
+ * The fluid node will next process after one full bucket rotation.
+ *
+ * Arguments:
+ * - [node][/obj/effect/particle_effect/fluid]: The node to start processing.
+ */
+/datum/controller/subsystem/fluids/proc/start_processing(obj/effect/particle_effect/fluid/node)
+	if (node.datum_flags & DF_ISPROCESSING || node.effect_bucket)
+		return
+
+	var/bucket_index = rand(1, num_effect_buckets) // Edit this value to make all fluids process effects (at the same time|offset by when they started processing| -> offset by a random amount <- )
+	effect_carousel[bucket_index] += node
+	node.effect_bucket = bucket_index
+	node.datum_flags |= DF_ISPROCESSING
+
+/**
+ * Stops processing the effects of a fluid node.
+ *
+ * Arguments:
+ * - [node][/obj/effect/particle_effect/fluid]: The node to stop processing.
+ */
+/datum/controller/subsystem/fluids/proc/stop_processing(obj/effect/particle_effect/fluid/node)
+	if(!(node.datum_flags & DF_ISPROCESSING))
+		return
+
+	var/bucket_index = node.effect_bucket
+	if(!bucket_index)
+		return
+
+	effect_carousel[bucket_index] -= node
+	if (bucket_index == effect_bucket_index)
+		currently_processing -= node
+
+	node.effect_bucket = null
+	node.datum_flags &= ~DF_ISPROCESSING
+
+#undef SS_PROCESSES_SPREADING
+#undef SS_PROCESSES_EFFECTS
+
+
+// Subtypes:
+
+/// The subsystem responsible for processing smoke propagation and effects.
+FLUID_SUBSYSTEM_DEF(smoke)
+	name = "Smoke"
+	spread_wait = 0.1 SECONDS
+	effect_wait = 2.0 SECONDS
+
+/// The subsystem responsible for processing foam propagation and effects.
+FLUID_SUBSYSTEM_DEF(foam)
+	name = "Foam"
+	spread_wait = 0.2 SECONDS
+	effect_wait = 0.2 SECONDS
```

Comment:
subsystem wait is in deciseconds, it might be better to speak in that directly, depending. up to you really

Reply:
Use of the seconds macros seems to be about 50:50 among current subsystems. I prefer them so I think I'll keep them.